### PR TITLE
partition: fix hash partition with not between condition get wrong result

### PR DESCRIPTION
**Pull request by xiongjiwei**
### What problem does this PR solve?

Issue Number: close https://github.com/pingcap/tidb/issues/22892

Tests 

- Unit test

### Release note 

- fix a bug that result of hash partition is wrong if where condition is `not between`

### CODE REVIEW

**Review by winoros — APPROVED** [review]

**Review by Reminiscent — APPROVED** [review]
/LGTM

**Comment by Reminiscent** [comment]
Should we cherry-pick this PR?

**Comment by xiongjiwei** [comment]
> Should we cherry-pick this PR?

no, it is not a critical issue

File: util/ranger/points.go [hunk 1]
```diff
@@ -638,6 +638,8 @@ func (r *builder) buildFromNot(expr *expression.ScalarFunction) []*point {
 		startPoint := &point{value: types.MinNotNullDatum(), start: true}
 		endPoint := &point{value: types.MaxValueDatum()}
 		return []*point{startPoint, endPoint}
+	case ast.LogicAnd:
+		return r.intersection(r.build(expr.GetArgs()[0]), r.build(expr.GetArgs()[1]))
```
Comment (wjhuang2016) [inline util/ranger/points.go:642]: WTF， ┐(a and b) => (a and b)? 
Did you forget that this function is buildFromNot() ?
@xiongjiwei @winoros @Reminiscent

**Pull request merged**